### ISSUE TITLE
Non-zero exit code when all linting succeeded

### ISSUE THREAD

**Issue by kpfleming** [ansible/ansible-lint#5147]
##### Summary

A similar flavor of this was reported earlier in the year, but I'm still experiencing it with version 26.8.0.

The issue is that ansible-lint runs against my large set of playbooks and roles, does not find any problems or rule violations, reports that the 'production' profile passed, but reports failure with a non-zero exit code.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
ansible-lint --version

ansible-lint 26.8.0 using ansible-core:2.21.3 ansible-compat:26.6.0 ruamel-yaml:0.19.1 ruamel-yaml-clib:None
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### STEPS TO REPRODUCE

I don't have a minimal test case to report, which I understand will complicate troubleshooting, but I'm happy to work with someone to do whatever is needed to obtain information which will be useful.

```console (paste below)
kpfleming@balrog24:~/git-personal/km6g/ansible$ ansible-lint --fix=none --strict --profile=production

Failed: 0 failure(s), 0 warning(s) in 752 files processed of 863 encountered. Profile 'production' was required, and it passed.
kpfleming@balrog24:~/git-personal/km6g/ansible$ echo $?
2
```

##### Desired Behavior

The exit code should be zero and no 'failure' should be reported.

##### Actual Behavior

https://gist.github.com/kpfleming/4c715000f6d85abae627eeb1aa97ca82

**Comment by arunrajm91** [comment]
Root-caused this, with a minimal reproduction. The trigger is `--fix=none` combined with `--strict` and at least one violation the auto-fixer could resolve (e.g. `yaml[empty-lines]`):

```console
$ printf -- '---\n- name: Test play\n  hosts: localhost\n  tasks:\n    - name: Print something\n      ansible.builtin.debug:\n        msg: hello\n\n\n' > playbook.yml
$ ansible-lint --fix=none --strict --profile=production playbook.yml; echo rc=$?
Failed: 0 failure(s), 0 warning(s) in 1 files processed of 1 encountered. Profile 'production' was required, and it passed.
rc=2
```

What happens, in order:

1. `--strict` decides failure early in `main()` because `result.matches` is non-empty at that point.
2. `--fix=none` makes `options.write_list == ["none"]`, which is truthy, so `fix()` runs anyway.
3. Inside the transformer, the FormattedYAML re-serialization is not gated by the `"none"` write set — so `--fix=none` actually **rewrites files on disk** (compare the file's hash before/after; the trailing blank lines above get removed).
4. The rerun loop in `fix()` then sees those matches resolved in the rewritten file and pops them from `result.matches` — so `count_results()` later reports 0 failures / 0 warnings and the profile as passed, while the exit code was already locked to 2.

This also explains the success-shaped numbers in your 752-of-863 run: the fixable matches were silently fixed on disk and dropped from the report *after* the strict decision. A second run on the same tree can then exit 0, because the files were modified by the first one.

PR incoming: `fix()` now returns early when the effective write set is `{"none"}`, so `--fix=none` behaves exactly like plain linting — files untouched, violations reported, exit code consistent with the report.

**Comment by kpfleming** [comment]
Thank you so much for this detailed report!
